### PR TITLE
[MSCRIPTING-7] binding the session and servers helper in binding context

### PR DESCRIPTION
**Pull request by rmannibucau**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MSCRIPTING) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MSCRIPTING-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSCRIPTING-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java [hunk 1]
```diff
@@ -73,8 +84,10 @@ public void execute()
          AbstractScriptEvaluator execute = constructExecute();
 
          Bindings bindings = new SimpleBindings();
+         bindings.put( "session", session );
          bindings.put( "project", project );
          bindings.put( "log", getLog() );
+         bindings.put( "servers", new Servers( session, settingsDecrypter ) );
```
Comment (rfscholte) [inline src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java:90]: I have doubts about this one, `servers` has never been used as a variable. Better to stick to `settings` with resolved values.

**Comment by rfscholte** [comment]
Don't forget to update https://github.com/apache/maven-scripting-plugin/blob/master/src/site/markdown/script-context.md

File: src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java (hunk 1 above)
Comment (rmannibucau) [reply src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java:90]: Do we have it already? I would like to avoid to decipher the whole settings each time (this is done only when needed) because it can slow down the execution a loooot for nothing in practise.
Happy to wrap it in a MavenScripting wrapper which would expose this kind of API (so script would use mavenScripting.servers.find(...))

Comment (rfscholte) [reply src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java:90]: If I understand what you need and staying close to what Maven offers, I think you should focus on `session.getRepositorySession().getAuthenticationSelector().getAuthentication(RemoteRepository)`

Comment (rmannibucau) [reply src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java:90]: What I want is a Server#password in clear but I don't want to decipher all servers upfront. From what I saw, getSettings() is NOT clear there so I would have to "clearify" it before binding it which I want to avoid so I thought using a lazy simple accessor is more script friendly than exposing the session and letting user use plexus container to do lookups to decipher values.

**Comment by rfscholte** [comment]
I'm going to close this one as I consider this a security vulnerabilty. The introduction of `servers` with exposed username/password means you can easily extract the credentials from the system it is running on.

**Comment by rmannibucau** [comment]
@rfscholte will you also prevent using any mojo? It is the same security level. Without that this plugin is not that helping in most cases IMHO (no easy website custo/generation), no CI/CD etc...
Worked around this lack by using exec plugin and a custom maven deceypter - it is just 100LoC but the security point is clearly not accurate for any mojo since they all have access to it and they are all - script or not - part of the build. Preventing a script not in the project is a thing but not all other cases technically speaking.

**Comment by bmarwell** [comment]
You could also add:

```
         bindings.put( "pluginDescriptor", pluginDescriptor );
         bindings.put( "mojoExecution", mojoExecution );
```

from #7  if you think it adds any benefits.
Needs a rebase and `spotless:apply` either way.

**Comment by rmannibucau** [comment]
@bmarwell this is true but this is also something the script somehow know in the context of a project. What would be more useful is a replacement for plexus container to lookup any potentially needed bean or a loose coupling like in exec:java (constructor injection runnable case) but my suspicion is maven (4) misses a container handler for these cases.

**Comment by bmarwell** [comment]
Re-opened for new evaluation

**Comment by cstamas** [comment]
A big -1 here

**Comment by rmannibucau** [comment]
@cstamas any rational behind and proposal to solve the related issue (enable to do proper resolutions from the script)?

**Comment by cstamas** [comment]
This is a matter of principle, and IMHO, vision.

Plugins like these are like "swiss knives" and allows hacking, in effect, like Ant or other tools allow it (and makes Maven hackable just line one would do with Ant). 

Maven IMHO should be more like "set of Un*x tools" instead: set of battle tested (but also simple) plugins, that do few things, but do them well. 

If someone is akin for "swiss knife" plugins like these, for me is clear sign that they want to even widen their "hacking", that is IMHO another sign, they should be using Gradle, Ant or whatever else. Maven is a _declarative build system_.

And I did not even mention the reasons Robert brought up.... (plus multitude of other reasons).

**Comment by bmarwell** [comment]
> This is a matter of principle, and IMHO, vision.
> 
> Plugins like these are like "swiss knives" and allows hacking, in effect, like Ant or other tools allow it (and makes Maven hackable just like one would do with Ant).
> 
> Maven IMHO should be more like "set of Un*x tools" instead: set of battle tested (but also simple) plugins, that do few things, but do them well.
> 
> If someone is akin for "swiss knife" plugins like these, for me is clear sign that they want to even widen their "hacking", that is IMHO another sign, they should be using Gradle, Ant or whatever else. Maven is a _declarative build system_.
>
> And I did not even mention the reasons Robert brought up.... (plus multitude of other reasons).

 This is also already possible in the exec plugin, and no one said this was a vuln (see Romains comment). Then, this would mean you need also get rid of exec-plugin.

Then, what about the requirements history?

Why was this plugin released in the first place? There could have been a vote against it at that time?

**Comment by cstamas** [comment]
It is not all black or white of course.

We do need some "backdoors" open, for case of "emergency", if no plugin exists that does what you need. But if you google it, you will see the trend: "use hack in maven, but if you need it too often, write a plugin for it" tendency. And we must not lose this.

This is why originally writing maven plugins was possible using ant scripts and beanshell as well. Beanshell, Ant script, they were all there to lower the plugin authoring barrier, BUT, those were plugins, they used plugin API, had descriptors, etc, so any future thing like build avoidance, incremental build (real one, not what we have today), etc. could rely on them.

Problem is when "hacks" remain hacks, or worse, when "hacks become pattern" (of using some "special" (backdoor) plugin). So if you have ant/exec plugin in build, it is okay, as long as you do not stick with it. But in our case, I see pattern of building some "library of hacks" on top of these, that makes ANY of these "solutions" non reusable, non shared, not maven plugins 

If you need something special, hack it once. But second time write a plugin for it. Period.

In short, we do need "stepping stone" plugins (hacks, like exec, ant plugins) to solve some ad-hoc scenarios, BUT these are stepping stones, in a sense, if pattern is repeated, they should be made into reusable plugins ultimately, while I see no intention of that happening here. In contrary, I see quite the opposite intent.

**Comment by rmannibucau** [comment]
@cstamas I'm on the camp to enable people to do what works for them, for me maven key is dependencies and it must stay static, rest (build pipeline) already broke maven original design and convention and it is bad to need to create a module/project for a build need so I'm clearly keen to see this kind of thing happening. No strong push from me there since I moved to exec and some companions but think it is sane.

**Comment by bmarwell** [comment]
I'm somewhere between the lines here. If you have a common use case, a dedicated plugin is probably a better solution compared to the exec-, antrun-, or scripting-plugin.

I'll say let's skip this pr for now... There's also #7 FWIW. 
If more people need such a backdoor, we can visit this and #7 again.

File: src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java [hunk 2]
```diff
@@ -60,18 +64,27 @@ public class EvalMojo extends AbstractMojo {
...
 
+    @Component
+    private SettingsDecrypter settingsDecrypter;
+
     // script variables
     @Parameter(defaultValue = "${project}", readonly = true)
     private MavenProject project;
 
+    // script variables
+    @Parameter(defaultValue = "${session}", readonly = true)
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java:75]: @Component can replace read-only parameter

Comment (rmannibucau) [reply src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java:75]: Is there a pro? I tend to prefer to limit the API (#imports) but if there is some advantage i'll do

**Comment by jira-importer** [comment]
Resolve #44

**Comment by elharo** [comment]
Tests are failing. Fixing that would be required for review. I don't have a particular opinion on this, but I note that other people whose opinion I respect are all negative about this, so I don't hold out high hopes here. It's uncommon for a PR to get multiple informed -1 votes so it's probably worth paying attention to.

File: src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java (hunk 2 above)
Comment (slachiewicz) [reply src/main/java/org/apache/maven/plugins/scripting/EvalMojo.java]: Simply change to @Inject

File: src/test/java/org/apache/maven/plugins/scripting/binding/ServersTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import org.apache.maven.settings.Proxy;
+import org.apache.maven.settings.Server;
+import org.apache.maven.settings.Settings;
+import org.apache.maven.settings.building.SettingsProblem;
+import org.apache.maven.settings.crypto.DefaultSettingsDecrypter;
+import org.apache.maven.settings.crypto.SettingsDecryptionRequest;
+import org.apache.maven.settings.crypto.SettingsDecryptionResult;
+import org.codehaus.plexus.PlexusContainerException;
+import org.junit.jupiter.api.Test;
+import org.sonatype.plexus.components.cipher.DefaultPlexusCipher;
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/scripting/binding/ServersTest.java]: why are we using org.sonatype instead of org.codehaus.plexus?

**Comment by DavoBR** [comment]
Hi there! I’ve been following this PR and I think exposing the MavenSession instance is a crucial addition for this plugin. It would significantly increase the scripts' capabilities by allowing access to project properties, settings, and the reactor state in a straightforward way.

I personally prefer this plugin over others like gmavenplus due to its simplicity and 'Maven-native' feel. Is there anything 

I can help with to move this forward? It would be great to see this merged and released. Thanks for the effort!

File: src/test/java/org/apache/maven/plugins/scripting/binding/ServersTest.java (hunk 3 above)
Comment (rmannibucau) [reply src/test/java/org/apache/maven/plugins/scripting/binding/ServersTest.java:32]: from memory it was cause it is the one in the stack so kept the dep management simple and since it was not critical for tests it was a good compromise IMO

**Comment by rmannibucau** [comment]
I'm a bit lost in review, while there is still a strong need of this feature and it doesnt bring more security flaws than already existing (acp it is with acks but trivial acks today) I think we should move it forward but I'm not sure if we are good or need some action.

**Comment by rmannibucau** [comment]
up? got bitten again twice in 2 weeks by that and can't embrace scripting without a lot of reflection (to not cite the worse of the hacks) cause of the lack of session availability, since there is no (more) security issue (than in default maven pipeline) can we move forward and integrate this trivial change?

File: src/test/java/org/apache/maven/plugins/scripting/binding/ServersTest.java (hunk 3 above)
Comment (elharo) [reply src/test/java/org/apache/maven/plugins/scripting/binding/ServersTest.java:32]: I don't see any other import of org.sonatype.plexus in this project. I also don;t see it added to a pom.xml in this PR. I guess it's coming from a transitive dependency? In any case, let's not ad it here as long as the regular codehaus version works.

**Review by elharo — CHANGES_REQUESTED** [review]

**Comment by rmannibucau** [comment]
>  In any case, let's not ad it here as long as the regular codehaus version works.

well it works but it is not there AFAIK whereas this one is in the testing stack so I prefered to avoid to pull yet another dep for no real reason, this is why it is there
